File: services/billing-cdc-worker/src/main/java/com/openimago/billingcdc/handler/models/BillingEvent.java
package com.openimago.billingcdc.handler.models;

import com.fasterxml.jackson.databind.JsonNode;

/**
 * A parsed Debezium change event for billing processing.
 *
 * @param sourceLsn  PostgreSQL Log Sequence Number from the event source
 * @param txId       PostgreSQL transaction ID
 * @param tableName  fully-qualified table name (e.g., {@code public.session})
 * @param operation  operation type: {@code r} (read/snapshot), {@code c} (create/insert),
 *                   {@code u} (update), {@code d} (delete)
 * @param primaryKey primary key value of the changed row
 * @param before     the row state before the change (null for inserts)
 * @param after      the row state after the change (null for deletes)
 */
public record BillingEvent(
        String sourceLsn,
        long txId,
        String tableName,
        String operation,
        String primaryKey,
        JsonNode before,
        JsonNode after
) {

    /**
     * Creates a uniqueness key from source metadata and operation details.
     *
     * <pre>{@code
     * source_lsn :: txid :: table_name :: operation :: primary_key
     * }</pre>
     */
    public String uniquenessKey() {
        return String.join("::",
                sourceLsn,
                String.valueOf(txId),
                tableName,
                operation,
                primaryKey
        );
    }

    /**
     * Returns true if this event is an INSERT.
     */
    public boolean isInsert() {
        return "c".equals(operation);
    }

    /**
     * Returns true if this event is an UPDATE.
     */
    public boolean isUpdate() {
        return "u".equals(operation);
    }

    /**
     * Returns true if this event is a DELETE.
     */
    public boolean isDelete() {
        return "d".equals(operation);
    }
}

File: packages/compute/model/billing_event.go
package model

import (
	"fmt"
	"math"
	"strings"
)

// BillingEvent is a parsed CDC change event for billing processing.
type BillingEvent struct {
	SourceLSN       string
	TxID            int64
	TableName       string
	Operation       string
	PrimaryKey      string
	TsMs            int64
	WorkspaceID     *string
	ProjectID       *string
	BeforeCost      *float64
	AfterCost       *float64
	TokensInput     *int64
	TokensOutput    *int64
	TokensReasoning *int64
	TokensCacheRead *int64
	TokensCacheWrite *int64
}

// UniquenessKey creates a uniqueness key from source metadata and operation details.
// Format: source_lsn :: txid :: table_name :: operation :: primary_key
func (e BillingEvent) UniquenessKey() string {
	return strings.Join([]string{
		e.SourceLSN,
		fmt.Sprintf("%d", e.TxID),
		e.TableName,
		e.Operation,
		e.PrimaryKey,
	}, "::")
}

// IsInsert returns true if this event is an INSERT.
func (e BillingEvent) IsInsert() bool { return e.Operation == "c" }

// IsUpdate returns true if this event is an UPDATE.
func (e BillingEvent) IsUpdate() bool { return e.Operation == "u" }

// IsDelete returns true if this event is a DELETE.
func (e BillingEvent) IsDelete() bool { return e.Operation == "d" }

// IsSnapshot returns true if this is a snapshot/read event (should be skipped).
func (e BillingEvent) IsSnapshot() bool { return e.Operation == "r" }

// DeltaMicros computes the cost delta in micros (after - before, rounded).
// Returns 0 if costs are nil/missing.
func (e BillingEvent) DeltaMicros() int64 {
	if e.BeforeCost == nil || e.AfterCost == nil {
		return 0
	}
	delta := *e.AfterCost - *e.BeforeCost
	return int64(math.Round(delta * 1_000_000.0))
}
